handler/rfc8628/token_handler.go

```diff
@@ -6,69 +6,99 @@ package rfc8628
 import (
 	"context"
 
+	"github.com/pkg/errors"
+
 	"github.com/ory/x/errorsx"
 
 	"github.com/ory/fosite"
 	"github.com/ory/fosite/handler/oauth2"
 )
 
-// DeviceTokenHandler is a token response handler for the Device Code introduced in the Device Authorize Grant
-// as defined in https://www.rfc-editor.org/rfc/rfc8628
-type DeviceTokenHandler struct {
+type DeviceCodeHandler struct {
 	DeviceRateLimitStrategy DeviceRateLimitStrategy
 	DeviceCodeStrategy      DeviceCodeStrategy
-	DeviceCodeStorage       DeviceCodeStorage
 }
 
-func (c DeviceTokenHandler) ValidateGrantTypes(ctx context.Context, requester fosite.AccessRequester) error {
-	if !requester.GetClient().GetGrantTypes().Has(string(fosite.GrantTypeDeviceCode)) {
-		return errorsx.WithStack(fosite.ErrUnauthorizedClient.WithHint("The OAuth 2.0 Client is not allowed to use authorization grant \"urn:ietf:params:oauth:grant-type:device_code\"."))
+func (c DeviceCodeHandler) Code(ctx context.Context, requester fosite.AccessRequester) (code string, signature string, err error) {
+	code = requester.GetRequestForm().Get("device_code")
+
+	if c.DeviceRateLimitStrategy.ShouldRateLimit(ctx, code) {
+		return "", "", errorsx.WithStack(fosite.ErrPollingRateLimited)
 	}
 
-	return nil
+	signature, err = c.DeviceCodeStrategy.DeviceCodeSignature(ctx, code)
+	if err != nil {
+		return "", "", errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
+	}
+
+	return
 }
 
-func (c DeviceTokenHandler) ValidateCode(ctx context.Context, requester fosite.AccessRequester, code string) error {
+func (c DeviceCodeHandler) ValidateCode(ctx context.Context, requester fosite.AccessRequester, code string) error {
 	return c.DeviceCodeStrategy.ValidateDeviceCode(ctx, requester, code)
 }
 
-func (c DeviceTokenHandler) GetCodeAndSession(ctx context.Context, requester fosite.AccessRequester) (code string, signature string, authorizeRequest fosite.Requester, err error) {
-	code = requester.GetRequestForm().Get("device_code")
+type DeviceSessionHandler struct {
+	DeviceCodeStorage DeviceCodeStorage
+}
 
-	if c.DeviceRateLimitStrategy.ShouldRateLimit(ctx, code) {
-		return "", "", nil, fosite.ErrPollingRateLimited
+func (s DeviceSessionHandler) Session(ctx context.Context, requester fosite.AccessRequester, codeSignature string) (fosite.Requester, error) {
+	req, err := s.DeviceCodeStorage.GetDeviceCodeSession(ctx, codeSignature, requester.GetSession())
+
+	if err != nil && errors.Is(err, fosite.ErrInvalidatedDeviceCode) {
+		if req == nil {
+			return req, fosite.ErrServerError.
+				WithHint("Misconfigured code lead to an error that prohibited the OAuth 2.0 Framework from processing this request.").
+				WithDebug("\"GetDeviceCodeSession\" must return a value for \"fosite.Requester\" when returning \"ErrInvalidatedDeviceCode\".")
+		}
+
+		return req, nil
```

Comment:
this is a personal preference so feel free to discard


```suggestion
		if req != nil {
		    return req, nil
		}

		return req, fosite.ErrServerError.
				WithHint("Misconfigured code lead to an error that prohibited the OAuth 2.0 Framework from processing this request.").
				WithDebug("\"GetDeviceCodeSession\" must return a value for \"fosite.Requester\" when returning \"ErrInvalidatedDeviceCode\".")
		}
```

i find it way easier to read when the simple case is the first and the complex is the last, same comment applies on some of the code above

not a biggie, happy to keep it as is